Feed list impl

karllindmark opened the pull request:
Implementation of display of feed.

What it does right now:
- Shows all the different feed events and their sub-data

What it does not support:
- Posting to feed (AsyncTaskLoader vs AsyncTask - which one should I go for with the POST-requests?)
- Viewing correct gravatars  - have we decided upon using Picasso for this?
- Hooah/Comment

You can start reviewing the current code base @peter-budo, and then I'll add commits for the "what it does not support" when you've poked me in the correct direction there. :horse:

Commit:
Add keys for Profiles

Commit:
Add the url for the feed resources

Commit:
Add more game platforms to looup by ID

Commit:
Add white container style to the layouts

Commit:
Add data for the social section (userId & so on)

Commit:
Create feed layout map that maps events to resources

Commit:
Create a Dummy adapter for mockup purposes

Commit:
Add JSON models for the Battle Feed

Commit:
[WIP] Get Feed list fragment to fetch data

Commit:
Refactor LoadingListFragment

Commit:
Add new JSON models for the Feed. More to come.

Commit:
[WIP] Creating FeedEventFactory for the feed items

Commit:
Refactor json models for the Battle Feed

Commit:
[WIP] Adjust UI to work with new data formats

Commit:
Renamed EventDAta to BaseEvent

Commit:
Update resources

Commit:
Alter the resource maps

Commit:
Update UrlFactory

Commit:
Add JSON models for the events

Commit:
Add generators for the events

Commit:
Add event type and factory

Commit:
Update Battle Feed fragment to use all the changes

Commit:
Add request data to ConnectionRequest

Commit:
Rebase change in dev into feed-list-impl

Commit:
Add image resources for Battlepacks and weapon accessories

Commit:
Remove ugly heading from list items

Commit:
Add string for Battlereport icon

Commit:
Updating resources (temporarily bundling both black & white images in the drawables

Commit:
Add static layout for BattlePackEvent

Commit:
Make sure placeholder for BattlePackEvent is visible in the feed

Commit:
Add color to the accessory title

Commit:
Change LinearLayout to TableRow in GameReport view

Commit:
Remove old files not being used anymore

Commit:
Add camo images to the project, but keep them out of R.java

Add camo images to the project right now, but keep them out of R.java
until the names are correct. *sigh*

Commit:
Add XP boost images

Commit:
Add more resource maps to the repo

Commit:
[WIP] Add Battle Pack functionality to the feed

Commit:
Rename camo images to correct name

Commit:
Move images to drawable top level

Commit:
Switching around the username/username2 IDs

Commit:
Update images, add missing ones

Commit:
Add content into the Views for Wall posts

Commit:
Add Camoflague image map

Commit:
Add String map for camoflagues

Commit:
Add emblems to repo, with both string- and image map

Commit:
Move the resource maps to categorized packages

Commit:
Update resource name

Commit:
Initial version of BattlePackEvent is now working

However, needs to be refactored before merging into dev

Commit:
Remove abstract class from BaseEvent

Commit:
Create interface and ui binders for the events

Commit:
Update UI to use new interface + renames

Commit:
Refactor of feed (ui binders)

Commit:
Add prefix to award images

Commit:
Add more maps to the project

Commit:
Add SHAREDGAMEEVENT event to the feed

Add SHAREDGAMEEVENT event to the feed. What a hassle!

Commit:
Update GameReportUiBinder to prevent NPE

A NPE could be raised if the soldier didn't have any "highlight" during
a given game (thus, getPlayerHightlight() would return null and the
subsequent call in the chain would cause the NPE).

Commit:
Ooops, forgot to implement FavoriteServerEvent

Oops, forgot to implement FavoriteServerEvent 100% - now it's displaying
the *real* server names in the feed. :))

Commit:
Refactor BattlePack* to use the new resource maps

Commit:
[WIP] Not sure if we need to pass this here

Not sure if we need to pass this here just yet - we'll leave it
commented a few commits more

Commit:
Update to use the new VehicleStringMap

Commit:
Refactor #0

Pre push refactor

Commit:
Extract strings in BattleFeedPostingFragment.java

peter-budo commented on BF4Intel/src/main/assets/battlefeed.json:
```diff
@@ -0,0 +1 @@
+{"type":"success","message":"OK","data":{"asInfo":{},"sessionUserId":"2832658801548551060","nowTimestamp":1385415193,"feedEvents":[{"numLikes":0,"owner":{"username":"CapTNRiggz","gravatarMd5":"3b2e81c77cebae2a29c0a0d98873f1d9","userId":"2832659598252935740","createdAt":1319515461,"presence":{"userId":"2832659598252935740","updatedAt":1385415193,"presenceStates":"0"}},"feedCategory":4,"personaId": ...
```
Move it to core resources where other json files are stored, I just wonder why we didn't do it with that other json file in assets too :confused:

Commit:
Move json files to core

Move json files to appropriate dir in core

peter-budo commented:
After this pull request and the one I'm working on, we will start pushing large collection of images beforehand as it does make a lot of distraction. Doesn't give correct feeling of what actual Java classes and Android layout and other important resources been altered. Beside it is my opinion this should have been worked on small chunks as it is hell way to much to approve, and big possibility of things been overlooked

Commit:
Remove unused attribute

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/json/battlefeed/uibinders/SharedGameEventUiBinder.java:
```diff
@@ -0,0 +1,102 @@
+                } else if (category.equals("BF4DOGTAGS")) {
+                    populateCell(cell, R.drawable.test_dogtag, DogtagStringMap.get(itemKey));
+                } else {
+                    Log.d(Battlelog.TAG, "[populateView] Unable to populate view for '" + category + "'");
```
Use `SharedGameEventUiBinder.class.getSimpleName()` instead of generic TAG

karllindmark replied:
Would getClass().getSimpleName() produce the same result?

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/battlefeed/BattleFeedFragment.java:
```diff
@@ -0,0 +1,154 @@
+
+    @Override
+    protected void onLoadFailure(final String resultMessage) {
+        Log.e(Battlelog.TAG, "resultMessage => " + resultMessage);
```
Use `SharedGameEventUiBinder.class.getSimpleName()` instead of generic TAG

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/Battlelog.java:
```diff
@@ -3,4 +3,5 @@
 import android.app.Application;
 
 public class Battlelog extends Application {
+    public final static String TAG = "com.ninetwozero.bf4intel";
```
Obsolete, see SharedGameEventUiBinder and BattleFeedFragment for reason and remove it please

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/ui/BaseLoadingListFragment.java:
```diff
@@ -48,23 +58,27 @@ public boolean onOptionsItemSelected(final MenuItem item) {
 
-    protected <T extends Object> T fromJson(final String jsonString, final Class<T> outClass) {
-        final JsonObject jsonObject = parser.parse(jsonString).getAsJsonObject().getAsJsonObject("data");
+    protected <T extends Object> T fromJson(final String json, final Class<T> outClass) {
```
Change class declaration from
`public abstract class BaseLoadingListFragment extends BaseListFragment implements LoaderManager.LoaderCallbacks<Result>`
to 
`public abstract class BaseLoadingListFragment<T> extends BaseListFragment implements LoaderManager.LoaderCallbacks<Result>`

where class extending this fragment will look like 
`public class BattleFeedFragment extends BaseLoadingListFragment<FeedItem>`

and then you can have methods as
`protected T fromJson(final String json, final Class<T> outClass)`

karllindmark replied:
I actually felt like I wanted to keep it on method level, as we may be doing multiple requests in the same fragment. Does that seem reasonable? We may not _only_ be getting ClassA, but maybe also ClassB.

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/ui/BaseLoadingListFragment.java:
```diff
@@ -48,23 +58,27 @@ public boolean onOptionsItemSelected(final MenuItem item) {
     }
 
-    protected <T extends Object> List<T> fromJsonArray(final String jsonString, final Class<T> outClass, final String container) {
+    protected <T extends Object> List<T> fromJsonArray(final Gson gsonToUse, final String json, final Class<T> outClass, final String container) {
```
Any specific reason why you need this, given that collection can be fetch from "parent" json class? (Parent class meant as the top class of JSON extraction, like Awards or Assignments that holds inside collection Award and Assignment)

karllindmark replied:
I wanted to keep away from using a parent class to wrap a single List, as it does feel a bit wrong. Not sure which is "better" though!

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/datatypes/EventType.java:
```diff
@@ -0,0 +1,30 @@
+import com.ninetwozero.bf4intel.R;
+
+public enum EventType {
+    STATUSMESSAGE(R.layout.list_item_feed_status),
```
I would have preferred names along the line of `STATUS_MESSAGE` which are easier to read unlike `CREATEDFORUMTHREAD`

karllindmark replied:
Good point!

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/factories/FragmentFactory.java:
```diff
@@ -5,7 +5,20 @@
 import com.ninetwozero.bf4intel.ui.battlereport.BattleReportFragment;
 import com.ninetwozero.bf4intel.ui.battlereport.BattleReportListingFragment;
-import com.ninetwozero.bf4intel.ui.fragments.*;
+import com.ninetwozero.bf4intel.ui.fragments.AccountProfileFragment;
```
Why you trying to go against IDE optimize imports? Mac `Ctrl+Alt+O`

karllindmark replied:
It's actually the IDE that's doing that "auto import" thingy.

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/factories/UiBinderFactory.java:
```diff
@@ -0,0 +1,47 @@
+    };
+
+    @SuppressWarnings("unchecked")
+    public static <T extends EventUiBinder> T get(final EventType eventType) {
```
Use of reflexion as per BaseLoadingListFragment.java

karllindmark replied:
Not sure I understand this comment (apart from it being related to BaseLoadingListFragment.java comment above)

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/json/battlefeed/BaseEvent.java:
```diff
@@ -0,0 +1,15 @@
+package com.ninetwozero.bf4intel.json.battlefeed;
```
Why this and other classes that are pure Java are not in core?

karllindmark replied:
I wanted to wait and see where we should put the UiBinders --> in the ui/battlefeed package?  :)

Commit:
Rename EventType fields for readability

Commit:
Optimize imports

Commit:
Remove Battlelog.TAG from the project

peter-budo commented:
This need more code management

> json.battlefeed.events: 
> Contains the POJO's for the different event types (GAMEACCESS, STATUSMESSAGE, BATTLEPACK and so on)

Datatypes inside json.battlefeed.events are actual JSON objets therefore should be inside json of core module.
Your events are POJO's and have no need of Android so can be also moved to core you may create pojo package on same level as json there

> json.battlefeed.generators:
> Contains the code needed to generate the *Event instance (ie, generates instance of Foo from json)

I do not follow why these "generators" are need it/necessary when you already parsed JSON to some Java object and you can then construct new POJO's with getters of that object

> json.battlefeed.uibinders:
> Constains the code needed to display the data an *Event contains in a given view (given view = part of loaded list layout for the Feed Item)

Need to be moved as current location is not ui section and they do bellog to UI. Maybe a view or layout postfix instead of uibinder would be more appropriate/comprehending?

Commit:
GSON:ify the events

GSON:ify the events by using gson on the whole object and setting the
eventType field in BaseEvent after the object has been created.

Commit:
Move json related stuff (non Android) to core

Move json related stuff (non Android) to core package. Alter classes to
make the transition OK.

peter-budo commented:
Almost there...

Core
- Can we do something about battlefeed `datatypes` and `misc` they are such ugly non-descriptive names? EventType should join battlefeed classes as it is linked with them. As for misc either rename it to deserializer or something that gives package more purposeful name
- Can't we move classes inside `datatypes` part of `battlefeed/events` to just `battlefeed` as is BaseEvent and FeedItem? I see no reason to burry them so low in structure

BF4Intel
- Newly added or currently edited layout have hardcoded text in TextViews. Can you please remove it? It is OK to use it while creating/updating but after it should be removed as it out lived it purpose and Lint does hate us for it

After this we are good to merge

karllindmark commented:
This feed item need to be handled:

```
"SHAREDGAMEEVENT":{"BF4TRACKINGCOMPLETE":{"unlocks":[],"assignments":[{"nameSID":"WARSAW_ID_P_XP1_AWARD_AS_08","unlockId":"xp1as08","imageConfig":{"category":"award_mission","slug":"xp1as08","texture":"XP1\/Art\/Persistence\/Award\/Assigments\/battlelog\/xp1a08_BL","versions":{"mobile":{"path":"warsaw\/gamedata\/award_mission\/mobile\/xp1as08.png","isSprite":false,"height":100,"name":"mobile","width":100},"large":{"path":"warsaw\/gamedata\/award_mission\/large\/xp1as08.png","isSprite":false,"height":256,"name":"large","width":256},"mediumns":{"path":"warsaw\/gamedata\/award_mission\/mediumns\/xp1as08.png","isSprite":false,"height":160,"name":"mediumns","width":160},"xsmall":{"path":"warsaw\/gamedata\/award_mission\/xsmall.png","isSprite":true,"height":43,"name":"xsmall","width":43},"small":{"path":"warsaw\/gamedata\/award_mission\/small.png","isSprite":true,"height":100,"name":"small","width":100},"smallns":{"path":"warsaw\/gamedata\/award_mission\/smallns\/xp1as08.png","isSprite":false,"height":100,"name":"smallns","width":100}}}}],"awards":[]}}
```

Commit:
Move EventType & co

Commit:
Refactor feed related items with Lint

Commit:
Adjust Battlefeed to include RANKUP ++

[X] Include BF4TRACKINGCOMPLETE subtype in Battle Feed
[X] Include BF4RANKUP subtype in Battle Feed
[X] Refactor SharedGameEvent (chopping into smaller pieces)
[X] Add missing expansion levels to map (XP1)
[X] Add missing "output" to BattlePackLayout (part of Battle Feed)
    We're now showing weapons as well (when they've been unlocked).
    However, we only have one image in that map due to the weapon
    statistics PR coming up. We need to make sure this gets connected
    when we're done with that one.

karllindmark commented:
@peter-budo: Alright buddy, I've updated the pull request with my "final" refactorings for this PR. Let me know if you deem it ok. To view the appropriate diffs, just click on 292efce, 922f6d8 and 8afa742 (as it's easy to lose track of things in "big diff").

Talk to you later! :sleeping:

peter-budo committed:
- Removed obsolete text in layouts
- Moved content descriptions strings from string.xml to content_string.xml
- Renamed misc_battlepack_strings.xml to battlepack_strings.xml
